### ISSUE TITLE
Handling exceptions in the REST API

### ISSUE THREAD

**Issue by rth** [FreeDiscovery/FreeDiscovery#31]
Currently when an exception happens in the REST API (e.g. wrong input arguments were provided), the following happens,
   - for all exceptions codes other than `HTTP 500`: a custom error message is returned `{''message': <custom_message>}`, defined in [`freediscovery/exceptions.py`](https://github.com/FreeDiscovery/FreeDiscovery/blob/master/freediscovery/exceptions.py#L29)  through the exception handlers in [`freediscovery/server/app.py`](https://github.com/FreeDiscovery/FreeDiscovery/blob/master/freediscovery/server/app.py#L79), which is OK.
   - the `HTTP 500` is a special case, and despite the presence of an exception handler in [`freediscovery/server/app.py`](https://github.com/FreeDiscovery/FreeDiscovery/blob/master/freediscovery/server/app.py#L73), this consistently returns `{"message": 'Internal Server Error'}` which is not very helpful (and should be fixed).

**Comment by DenisFLASH** [comment]
For the code 500, you mean [this line](https://github.com/FreeDiscovery/FreeDiscovery/blob/master/freediscovery/server/app.py#L76) ?

{'message': 'help'} ?

**Comment by rth** [comment]
Yes, that was just for testing (the normal output should be `error.to_dict()`), but neither of those is returned. A default message is printed instead (which means that our exceptions handler is not called for some reason). 

More on this is discussed [here](https://stackoverflow.com/questions/14993318/catching-a-500-server-error-in-flask),

> Default exception handling that kicks in when an exception occurs that is not caught. In debug mode the exception will be re-raised immediately, otherwise it is logged and the handler for a 500 internal server error is used. If no such handler exists, a default 500 internal server error message is displayed.

In our case the handler exists and we are not in "debug" mode; not sure why this is happening. That's not very critical, but it means that for 500 errors we can't get a meaningful error message from the API client side, unless we look at the server logs..

Also as far as I remember, when running unit tests we are in the "test" mode and the exception handling happens differently (so this might work with unit tests but not in the real word), which makes the debugging a bit more tricky.

**Comment by DenisFLASH** [comment]
hm, sounds tricky indeed. Those "debug" or "test" modes, how are they defined? In other words, what is it that says that we are in one or another mode? For tests, i guess it can be due to @pytest decorator or something like that. And "debug" or other, does it have something to do with versionfile, packaging procedure in setup.py etc.?

**Comment by rth** [comment]
Normally the test mode is defined by  `app.config['TESTING'] = True` see [flask docs](http://flask.pocoo.org/docs/0.11/testing/) and the debug mode also as `app.config['DEBUG'] = True` (or as runtime parameter of `app.run(..)`). They are flask specific, but maybe I missed something...

**Comment by DenisFLASH** [comment]
Got it, so it's mostly Flask-related conf. Thanks

**Comment by rth** [comment]
This issue might have been solved in PR #81, but it needs to be checked..
